packages/components/src/date-time/date/index.tsx

```diff
@@ -100,139 +61,255 @@ function DatePickerDay( { day, events = [] }: DatePickerDayProps ) {
+									isInvalidDate ? isInvalidDate( day ) : false
+								}
+								numEvents={
+									events.filter( ( event ) =>
+										isSameDay( event.date, day )
+									).length
+								}
+								onClick={ () => {
+									setSelected( [ day ] );
+									setFocusable( day );
+									onChange?.(
+										format(
+											// Don't change the selected date's time fields.
+											new Date(
+												day.getFullYear(),
+												day.getMonth(),
+												day.getDate(),
+												date.getHours(),
+												date.getMinutes(),
+												date.getSeconds(),
+												date.getMilliseconds()
+											),
+											TIMEZONELESS_FORMAT
+										)
+									);
+								} }
+								onKeyDown={ ( event ) => {
+									let nextFocusable;
+									if ( event.key === 'ArrowLeft' ) {
+										nextFocusable = subDays( day, 1 );
+									}
+									if ( event.key === 'ArrowRight' ) {
+										nextFocusable = addDays( day, 1 );
+									}
+									if ( event.key === 'ArrowUp' ) {
+										nextFocusable = subWeeks( day, 1 );
+									}
+									if ( event.key === 'ArrowDown' ) {
+										nextFocusable = addWeeks( day, 1 );
+									}
+									if ( nextFocusable ) {
+										setFocusable( nextFocusable );
+										if (
+											! isSameMonth(
+												nextFocusable,
+												viewing
+											)
+										) {
+											setViewing( nextFocusable );
+											onMonthPreviewed?.(
+												format(
+													nextFocusable,
+													TIMEZONELESS_FORMAT
+												)
+											);
+										}
+									}
+								} }
+							/>
+						);
+					} )
+				) }
+			</Calendar>
+		</div>
+	);
+}
 
-		onChange?.( newDate.set( momentTime ).format( TIMEZONELESS_FORMAT ) );
+type DayProps = {
+	day: Date;
+	column: number;
+	isSelected: boolean;
+	isFocusable: boolean;
+	isFocusAllowed: boolean;
+	isToday: boolean;
+	numEvents: number;
+	isInvalid: boolean;
+	onClick: () => void;
+	onKeyDown: ( event: KeyboardEvent ) => void;
+};
 
-		// Keep focus on the date picker.
-		keepFocusInside();
-	};
+function Day( {
+	day,
+	column,
+	isSelected,
+	isFocusable,
+	isFocusAllowed,
+	isToday,
+	isInvalid,
+	numEvents,
+	onClick,
+	onKeyDown,
+}: DayProps ) {
+	const ref = useRef< HTMLButtonElement >();
 
-	const getEventsPerDay = ( day: Moment ) => {
-		if ( ! events?.length ) {
-			return [];
+	// Focus the button when an arrow key is pressed, but only when focus was
+	// already within the calendar grid. This prevents stealing focus from an
+	// external field e.g. one of the TimePicker fields.
+	useEffect( () => {
+		if ( ref.current && isFocusable && isFocusAllowed ) {
+			ref.current.focus();
 		}
+	}, [ isFocusable ] );
+
+	return (
+		<DayButton
+			ref={ ref }
+			className="components-datetime__date__day" // Unused, for backwards compatibility.
+			disabled={ isInvalid }
+			tabIndex={ isFocusable ? 0 : -1 }
+			aria-label={ getDayLabel( day, isSelected, numEvents ) }
```

Comment:
`getDayLabel` looks like a candidate for `useMemo`.

Reply:
Agreed with you but then I ended up having to remove the `push` stuff from `getDayLabel` because it didn't really work in RTL languages. This means `getDayLabel` is now a simple `if` and therefore I think not necessary to memoize.